### PR TITLE
Arcade layout

### PR DESCRIPTION
**Pull request by tmat**
## Build output (/artifacts)

The layout of build output is documented in Arcade docs and is the same for all repositories on Arcade SDK. Some directories are optional, e.g. repositories that do not build VS components naturally do not create VSSetup and VSSetup.obj directories.
Repositories may put additional content under VSSetup/{config}, beyond what’s documented in the Arcade SDK. 
All builds outputs go under artifacts directory in the repository root.

### VSSetup directory
Roslyn adds the following to VSSetup:
-	DevDivPackages: packages inserted to VS CoreXT package feed and other files inserted to VS enlistment
-	Insertion\bootstrapper, Insertion\BranchInfo, Insertion\OptProf – files needed for training of partial NGEN optimization (aka OptProf, aka IBC). This will likely change soon as our OptProf infrastructure is still evolving.
-	Installer: The installer zip (Roslyn_Preview.zip) goes here 
-	VSIXes that are inserted to VS are placed under Insertion directory, those that are not inserted to VS are placed directly under VSSetup, e.g. RoslynDeployment.vsix.
-	Projects that build shipping VS components produce VSIXes of 2 flavors: 
o	“Experimental”: placed to VSSetup/{config} directory
o	“SystemComponent”: placed to VSSetup/{config}/Insertion

VSSetup directory is published by official build to Azure Pipelines Artifact Store and picked up from there by Roslyn Insertion Tool. Everything needed for insertion must be places to this directory.

## Engineering directory (/eng)

This directory contains scripts, targets and configuration used by the build.
The most interesting files for a Roslyn developer are:
-	cibuild.cmd/.sh
Runs CI build (both PR validation and official builds).
-	test-build-correctness.cmd
Validates consistency and correctness of the build.
-	test-determinism.cmd
Validates that our build is deterministic.
-	generate-compiler-code.cmd
Regenerates generated files used in the compilers (syntax trees, etc.).

You can run the above scripts directly from your local clone. No arguments are required, although you might want to specify configuration if Debug is not what you want to test. The scripts are designed to execute exactly the same logic as on CI machine, modulo real signing and publishing. These run in validation-only mode when executed in non-official build (locally or during PR validation).

## Common engineering directory (/eng/common)

This directory contains bootstrapper scripts that are common across all Arcade repositories. The Arcade engineering system will overwrite these periodically, so don’t make any permanent changes here, they will disappear without notice.

## Scripts directory (/scripts)

This directory is used for helper scripts that are manually executed and not essential for CI build.

## .dotnet, .tools, .packages

These directories might be created in the repository root under certain circumstances. The bootstrapper script uses them to store downloaded .NET Core (when the required version is not found on PATH), vswhere (when building from environment that is not initialized with VS env variables) and restored NuGet packages (on CI when not using global nuget cache), respectively.

## Official build artifacts publishing

The official build does not publish to cpvsbuild share anymore. Instead it publishes 
1)	VS Setup Components
These are VSIX packages that are used by VS installer built to artifacts/VSSetup/{config}/Insertion directory 
2)	CoreXT packages
Packages used by VS build – mapped to ExternalAPIs and Tools in Razzle
3)	Artifact VSSetup
Entire artifacts/VSSetup/{config} directory. 
4)	Artifact Packages
Entire artifacts/packages/{config} directory

Fixes https://github.com/dotnet/roslyn/issues/23547

### CODE REVIEW

File: eng/build.ps1 [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        Build-OptProfData
+    }
+
+    if ($testDesktop -or $testVsi -or $testIOperation) {
+        TestUsingOptimizedRunner
+    }
+
+    if ($launch) {
+        $devenvExe = Join-Path $env:VSINSTALLDIR 'Common7\IDE\devenv.exe'
+        &$devenvExe /rootSuffix RoslynDev
```
Comment (jaredpar) [inline eng/build.ps1]: > &$devenvExe /rootSuffix RoslynDev [](start = 7, length = 34)

Should use Exec-Process here as the exit code isn't validated.

**Comment by jaredpar** [comment]
> tar czf mono-${VERSION}.tar.gz mono

Why do we still have a scripts directory at all after this change?

---
Refers to: scripts/build_mono.sh:16 in 54a1ef2. [](commit_id = 54a1ef274b37ea1c9aa3499493eecb240c7d095e, deletion_comment = False)

File: eng/build.ps1 (hunk 1 above)
Comment (tmat) [reply eng/build.ps1]: Not sure why we have it but I'll update it to use `Exec-Process`

**Comment by tmat** [comment]
> Why do we still have a scripts directory at all after this change?

I moved all scripts that are not used in build/CI there. These can definitely use some clean up.

```
PublicApi
UploadAzureZip
GenerateSdkPackages
build_mono.sh
check-for-loc-changes.cmd
check-for-loc-changes.ps1
cleanup_perf.ps1
clean_benchview_tools.cmd
crossgen.sh
edit-designers.ps1
generate-badges.ps1
generate-compiler-code.cmd
generate-compiler-code.ps1
install_benchview_tools.cmd
README.md
Run-TestsWithProcessWatchdog.ps1
run_perf.ps1
sort-unshipped.cmd
test-nuget-cache.ps1
```

**Pull request merged**